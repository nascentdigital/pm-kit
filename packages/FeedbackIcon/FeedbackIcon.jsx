import React from 'react'
import { css } from '@emotion/core'
import PropTypes from 'prop-types'

// pm-kit
import { parkGreen } from '@pm-kit/colours'

// images
import { successPath, failPath } from '../../shared/svg/index'

// keys of available icons
export const FEEDBACK_ICON_NAME = {
  CHECKMARK: 'checkmark',
  ERROR: 'error',
  LOADER: 'loader',
}

// definitions of availabe svg icons
const FEEDBACK_ICON_DATA = [
  {
    key: FEEDBACK_ICON_NAME.CHECKMARK,
    file: successPath,
  },
  {
    key: FEEDBACK_ICON_NAME.ERROR,
    file: failPath,
  },
]

// keys of available sizes
export const SIZES = {
  SMALL: '24px',
  MEDIUM: '36px',
  LARGE: '48px',
}

const FeedbackIcon = ({ name, size, ...rest }) => {
  const iconData = FEEDBACK_ICON_DATA.filter((data) => data.key === name)

  const loader = css`
    @keyframes spinner {
      to {
        transform: rotate(360deg);
      }
    }
    border: 3px solid ${parkGreen};
    border-radius: 50%;
    border-right-color: transparent;
    display: inline-block;
    animation: 1s linear infinite spinner;
    animation-delay: var(--spinner-delay);
    height: ${size};
    width: ${size};
  `

  return (
    <>
      {name === 'loader' ? (
        <div css={loader} {...rest} />
      ) : (
        <>
          {iconData && iconData[0] && (
            <img alt={iconData[0].key} height={size} src={iconData[0].file} weight={size} {...rest} />
          )}
        </>
      )}
    </>
  )
}

FeedbackIcon.propTypes = {
  /**
   * The icon you want to render.
   */
  name: PropTypes.oneOf([FEEDBACK_ICON_NAME.CHECKMARK, FEEDBACK_ICON_NAME.ERROR, FEEDBACK_ICON_NAME.LOADER]).isRequired,
  /**
   * The height and width of feedback-icon.
   */
  size: PropTypes.oneOf([SIZES.SMALL, SIZES.MEDIUM, SIZES.LARGE]),
}

FeedbackIcon.defaultProps = {
  size: '1.125rem',
}

export default FeedbackIcon
