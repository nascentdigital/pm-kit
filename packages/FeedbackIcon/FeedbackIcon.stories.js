import React from 'react'
import FeedbackIcon, { FEEDBACK_ICON_NAME, SIZES } from './FeedbackIcon'
import { withKnobs, select } from '@storybook/addon-knobs'

export default {
  title: 'Design System|FeedbackIcon',
  component: FeedbackIcon,
  decorators: [withKnobs],
}

export const error = () => <FeedbackIcon name={FEEDBACK_ICON_NAME.ERROR} size={SIZES.SMALL} />

export const Playground = () => {
  return (
    <FeedbackIcon
      name={select(
        'name',
        [FEEDBACK_ICON_NAME.CHECKMARK, FEEDBACK_ICON_NAME.ERROR, FEEDBACK_ICON_NAME.LOADER],
        FEEDBACK_ICON_NAME.CHECKMARK
      )}
      size={select('size', [SIZES.SMALL, SIZES.MEDIUM, SIZES.LARGE], SIZES.SMALL)}
    />
  )
}

export const success = () => <FeedbackIcon name={FEEDBACK_ICON_NAME.CHECKMARK} size={SIZES.SMALL} />

export const waiting = () => <FeedbackIcon name={FEEDBACK_ICON_NAME.LOADER} size={SIZES.SMALL} />

Playground.story = {
  name: 'playground',
  parameters: { docs: { page: null, disable: true } },
}
