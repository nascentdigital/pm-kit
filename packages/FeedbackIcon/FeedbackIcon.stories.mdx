import { Meta, Story, Preview, Props, Subtitle } from '@storybook/addon-docs/blocks'
import FeedbackIcon from './FeedbackIcon.jsx'
import { version } from './package.json'

<Meta title="Design System|FeedbackIcon" componenent={FeedbackIcon} />

# Feedback Icon
<Subtitle>
version {version}
</Subtitle>

## Usage
* To install `FeedbackIcon` package with npm run terminal:

```jsx
npm i @pm-kit/feedback-icon
```
* Import `FeedbackIcon` in your componenent to render

```jsx
import FeedbackIcon from '@pm-kit/feedback-icon'
```

## Properties 

<Props of={FeedbackIcon} />

## Stories

#### Success
<Preview>
<Story name="Success">
  <FeedbackIcon state="success" />
</Story>
</Preview>

#### Error
<Preview>
<Story name="Error">
<FeedbackIcon state="error" />
</Story>
</Preview>

#### Waiting
<Preview>
<Story name="Waiting">
<FeedbackIcon state="waiting" />
</Story>
</Preview>

#### Disabled
<Preview>
<Story name="Disabled">
<FeedbackIcon state="disabled" />
</Story>
</Preview>