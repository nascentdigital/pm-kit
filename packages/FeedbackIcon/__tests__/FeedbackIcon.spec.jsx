import React from 'react'
import { render, mount } from 'enzyme'

import FeedbackIcon, { FEEDBACK_ICON_NAME, SIZES } from '../FeedbackIcon'

describe('FeedbackIcon', () => {
  const defaultProps = {
    name: FEEDBACK_ICON_NAME.CHECKMARK,
  }

  const doMount = (overrides = {}) => {
    const feedbackIcon = mount(<FeedbackIcon {...defaultProps} {...overrides} />)

    const findFeedbackIcon = () => feedbackIcon.find('img')

    return {
      feedbackIcon,
      name: feedbackIcon.find('name'),
      findFeedbackIcon,
    }
  }

  it('renders correct icon', () => {
    const checkMarkImage = doMount({ size: SIZES.SMALL }).findFeedbackIcon()
    expect(checkMarkImage).toHaveProp('alt', FEEDBACK_ICON_NAME.CHECKMARK)
  })
})
