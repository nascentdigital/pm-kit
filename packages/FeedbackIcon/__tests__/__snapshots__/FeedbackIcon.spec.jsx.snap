// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedbackIcon renders checkmark when success 1`] = `null`;

exports[`FeedbackIcon renders error icon when error 1`] = `null`;

exports[`FeedbackIcon renders nothing when disabled 1`] = `null`;

exports[`FeedbackIcon renders spinner when waiting 1`] = `null`;
